### ISSUE TITLE
Bug: [no-useless-assignment] False positive in code using TypeScript's `satisfies`

### ISSUE THREAD

**Issue by gebsh** [eslint/eslint#19002]
### Environment

Node version: v22.9.0
npm version: v10.8.3
Local ESLint version: v9.12.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 24.0.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```js
import { parser } from 'typescript-eslint';

export default [
	{
		files: ['**/*.mts'],
		languageOptions: {
			parser,
		},
		rules: {
			'no-useless-assignment': 'error',
		},
	},
];
```
</details>

```ts
const foo = 'foo';

if (true) {}

([foo] satisfies Array<string>).forEach(str => {
	console.log(str);
});
```

### What did you expect to happen?

The variable `foo` should not violate the `no-useless-assignment` rule.

### What actually happened?

The `no-useless-assignment` rule is reported for `foo`.

### Link to Minimal Reproducible Example

https://typescript-eslint.io/play/#ts=5.5.4&showAST=es&fileType=.ts&code=MYewdgzgLgBAZiEMC8MDkCRoNwChcCWcMAFFAE4CuApgJQwDeAvviQNqYC6MEAhlAQhwC1CDACC5crwCeAHmjkCYAOYA%2BWgDoE5AKK9gACxKKUaxrhgxQkEABtqmuyBUmKtPEw%2B4gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQHYHsBaWZRJZZQgQwoEsBzXAW0VwBd0pFpp9pJwYAL4ghQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I initially [reported this](https://github.com/typescript-eslint/typescript-eslint/issues/10112) to the typescript-eslint team as I can't reproduce this without using the TypeScript-specific `satisfies` operator, but their recommendation was to report a bug here.

As noted in that issue, removing `satisfies` or `if` makes the error disappear. The error is also present when the `if` statement is replaced with other statements that introduce a branch, for example `while (false) {}`.

**Comment by bradzacher** [comment]
The error also disappears if you switch the operator from `satisfies` to `as`. This is why I suggested it seems like a bug in the rule itself. I didn't see anything meaningfully different between the two ASTs or the scope analysis.

**Comment by fasttime** [comment]
Thanks for the issue @gebsh, I am able to reproduce this. In my understanding, this problem boils down to an irregularity in the visitor keys for the `TSSatisfiesExpression` node (i.e. the AST node that contains the `satisfies` operator). Specifically I'm running:

```js
import { visitorKeys } from "@typescript-eslint/visitor-keys";

console.log("TSTypeAssertion", visitorKeys["TSTypeAssertion"]);
console.log("TSSatisfiesExpression", visitorKeys["TSSatisfiesExpression"]);
```

and seeing this output:

```console
TSTypeAssertion [ 'expression', 'typeAnnotation' ]
TSSatisfiesExpression [ 'typeAnnotation', 'expression' ]
```

This causes the operands of `satisfies` to be traversed in reversed order: first RHS, then LHS. ESLint does not sort sibling nodes by location before traversing them (I'm not sure if it should). Anyway, the rule produces an incorrect result because it does not expect the node listeners to be called in that order.

I have two possible fixes in mind for this issue:
1. Go back to typescript-eslint and kindly ask them to flip the order of the visitor keys for `TSSatisfiesExpression`.
2. Make ESLint sort sibling nodes by start position before traversing them (in [lib/shared/traverser.js](https://github.com/eslint/eslint/blob/b3a08376cfb61275a7557d6d166b6116f36e5ac2/lib/shared/traverser.js#L141-L151)). I tried that and all unit tests were still passing, but I can't say if the current behavior is a bug or intentional but untested.

**Comment by fasttime** [comment]
@bradzacher what do you think?

**Comment by bradzacher** [comment]
> the rule produces an incorrect result because it does not expect the node listeners to be called in that order

Why does the visit order of a single node (that does no assignments) cause the rule to break?
It does seem odd that the visit order can cause weird behaviour like that.

---

The visitor keys were added in https://github.com/typescript-eslint/typescript-eslint/pull/5717 almost 2 years ago and I do not remember why it was added with a different key ordering...
We should probably look to correcting this.

**Comment by mdjermanovic** [comment]
> Why does the visit order of a single node (that does no assignments) cause the rule to break?

It breaks calculations of the range where identifiers are in a code path segment.

https://github.com/eslint/eslint/blob/064c8b612e2e4b773d6b25867f2045e3ceaa9d66/lib/rules/no-useless-assignment.js#L457-L466

https://github.com/eslint/eslint/blob/064c8b612e2e4b773d6b25867f2045e3ceaa9d66/lib/rules/no-useless-assignment.js#L212-L221

In this example, the range for the code path segment that represents `[foo] satisfies Array<string>).forEach` is calculated as the range from `Array` (because that's the first traversed identifier in the code path segment) to `forEach`, instead of from `foo` to `forEach`.

**Comment by mdjermanovic** [comment]
> removing `satisfies` or `if` makes the error disappear. The error is also present when the `if` statement is replaced with other statements that introduce a branch, for example `while (false) {}`.

Removing the branch makes `const foo = 'foo';` be in the same code path segment as `([foo] satisfies Array<string>).forEach`, so in this case the range starts from `foo` in `const foo` and the reference is in it.

**Comment by bradzacher** [comment]
Perhaps it would also be worth adding a guard to that `Identifier` selector to guard against such breakages from 3rd party parsers?

EG like

```js
if (!segmentInfo.first) {
  segmentInfo.first = node;
  segmentInfo.last = node;
} else {
  if (node.range[0] < segmentInfo.first.range[0]) {
    segmentInfo.first = node;
  }
  if (node.range[1] > segmentInfo.last.range[1]) {
    segmentInfo.last = node;
  }
}
```

**Comment by bradzacher** [comment]
My question would be "would it hurt to do both"?

**Comment by mdjermanovic** [comment]
There are other rules and parts of the core (code path analysis, as mentioned) that assume nodes will be visited in the order they appear in the code, which is the expected order so I'm not in favor of updating this rule. It helped to find a problem that may affect other rules.

**Comment by bradzacher** [comment]
It does seem weird to rely on one rule failing in a rather obscure way to catch a misconfiguration in 3rd party parsers but if you don't wanna change it that's entirely up to you.

We'll fix the keys regardless.

**Comment by mdjermanovic** [comment]
> It does seem weird to rely on one rule failing in a rather obscure way to catch a misconfiguration in 3rd party parsers

I didn't mean this should be motivation to keep the current code, but if we update this rule, then it would make sense to check the entire codebase and update all code that expects the correct order because we should consistently either rely on it or not. That is my opinion, but if others are in favor of updating this rule, I won't be opposed.

**Comment by fasttime** [comment]
I have to agree with @mdjermanovic. Since there is an expectation that nodes are visited in the order they appear in the code it should not be necessary to account for alternative orderings of visitor keys.

**Comment by fasttime** [comment]
> The visitor keys were added in [typescript-eslint/typescript-eslint#5717](https://github.com/typescript-eslint/typescript-eslint/pull/5717) almost 2 years ago and I do not remember why it was added with a different key ordering...

It seems problematic that that code has been around unnoticed for such a long time. Maybe ESLint should throw an error if nodes are visited in an unexpected order. This could be done for example example in `RuleTester` (where we [wrap the parser](https://github.com/eslint/eslint/blob/17a07fb548ecce24b88e8b2b07491c24ed1111a9/lib/rule-tester/rule-tester.js#L265-L294) to instrument the AST), or in some other dedicated utility. Thoughts?

**Comment by gebsh** [comment]
The latest release of `typescript-eslint` fixed this problem, so I'd like to close this issue. However, my understanding is that there's still a pending discussion about ESLint internals. Should I leave this issue open, or would it make more sense to extract that discussion to a separate issue given that the bug described here is already resolved?

**Comment by fasttime** [comment]
> The latest release of `typescript-eslint` fixed this problem, so I'd like to close this issue. However, my understanding is that there's still a pending discussion about ESLint internals. Should I leave this issue open, or would it make more sense to extract that discussion to a separate issue given that the bug described here is already resolved?

Thanks for the feedback @gebsh, good to know that the problem has been solved. I think we should leave this issue open until we decide how to follow-up. Thoughts on https://github.com/eslint/eslint/issues/19002#issuecomment-2406667251 @eslint/eslint-team?

**Comment by nzakas** [comment]
The point of visitorKeys is for a language/parser to tell ESLint the order in which nodes should be traversed regardless of location information. I don't think it's a safe assumption that all languages and all trees will always want nodes traversed in the order in which they appear in the source text, so I'm not in favor of updating how we use visitorKeys internally. It's up to the language/parser to define those correctly.
